File: UNCW2018/part2/codes/twodturb.jl
#= twodturb.jl

Solves the 2D vorticity equation using a Fourier pseudospectral method
in a doubly-periodic box. Time-stepping is Forward Euler. Does not
dealias.
=#

using PyPlot


# Parameters
nx = 64            # Resolution
Lx = 2.0*pi        # Physical box size
nu = 3e-4          # Kinematic viscosity
dt = 2e-2          # Time step
nsteps = 60000      # Number of time steps


# Construct the grid
ny, Ly = nx, Lx
dx, dy = Lx/nx, Ly/ny
nk, nl = Int(nx/2+1), ny

x = 0.0:dx:Lx-dx
y = 0.0:dy:Ly-dy
k = 2.0*pi/Lx*(0:nk)
l = 2.0*pi/Ly*cat(1, 0:nl/2, -nl/2+1:-1)

# 2D wavenumber grids with shapes (nk, nl)
K = [ k[i] for i in 1:nk, j in 1:nl ]
L = [ l[j] for i in 1:nk, j in 1:nl ]

# Need K^2 = k^2 + l^2 and 1/K^2.
Ksq = K.^2.0 + L.^2.0
invKsq = 1.0./Ksq
invKsq[1, 1] = 0.0  # Will eliminate 0th mode during inversion


# Random initial condition
t = 0.0
zeta = rand(nx, ny)
zetah = rfft(q)

u = v = zeros(nx, ny)
psih = rhs = zeros(nk, nl)


# Step forward
fig = figure(); tic()
for step = 1:nsteps

  if step % 1000 == 0 && step > 1
    zeta = irfft(zetah, nx)
    clf(); pcolormesh(x, y, zeta);
    xlabel(L"x")
    ylabel(L"y")
    title("\$\\zeta(x, y, t = $t)\$")
    pause(0.01)

    cfl = maximum(u)*dx/dt
    @printf("step: %04d, t: %6.1f, cfl: %.2f, ", step, t, cfl)
    toc(); tic()
  end

  # Calculate right hand side of vorticity equation.
  # (Note that this algorithm is not as memory efficient as it
  # could be because it creates many temporary arrays.)
  psih = -invKsq .* zetah

  zetax = irfft(im*K.*zetah, nx)
  zetay = irfft(im*L.*zetah, nx)
  u = -irfft(im.*L.*psih, nx)
  v =  irfft(im.*K.*psih, nx)

  rhs = -rfft( u.*zetax + v.*zetay) .- nu*Ksq.*zetah

  # Step forward
  zetah .+= dt.*rhs
  t += dt
end


# Final plot
zeta = irfft(zetah, nx)
clf()
pcolormesh(x, y, zeta)

colorbar()
xlabel(L"x")
ylabel(L"y")
title("\$\\zeta(x, y, t = $t)\$")

show()

File: UNCW2018/part2/codes/twodturb.m
clear all;


% Time step the 2d vorticity equation on a doubly-periodic
% domain using a pseudospectral method


% Parameters
nu = 3e-4;          % Viscosity
Lx = 2*pi;          % Domain size
nx = 64;           % Resolution
dt = 2e-2;          % Time step
nsteps = 60000;      % Number of time steps


% Construct square physical and wavenumber grid
dx = Lx/nx;
x = 0:dx:Lx-dx;
k = 2*pi/Lx * [ 0:nx/2, -nx/2+1:-1 ];

y=x; ny=nx; Ly=Lx; dy=dx; l=k;    % Square domain

[X, Y] = meshgrid(x, y);
[K, L] = meshgrid(k, l);

Ksq = K.^2 + L.^2;         % Square wavenumber
invKsq = 1./Ksq;           % Inverse square wavenumber
invKsq(1, 1) = 0;          % for vorticity inversion


% Initial condition
t = 0.0;
zeta = rand(nx, nx);
zetah = fft2(zeta);


% Step forward
fig = figure(1);
t1 = tic;
for step = 1:nsteps

    % Quick plot
    if mod(step, 1000) == 0 && step > 1
        disp([ ...
            'step = '      num2str(step,    '%04d'),   ', ', ...
            't = '         num2str(t,       '%06.1f'), ' s, ', ...
            'wall time = ' num2str(toc(t1), '%0.3f')]), t1=tic;
        zeta = real(ifft2(zetah));
        clf, pcolor(X, Y, zeta); shading flat, axis xy, axis square, pause(0.01)
    end


    % Calculate right hand side
    psih = -zetah .* invKsq;
    zetax = real(ifft2(1i*K.*zetah));
    zetay = real(ifft2(1i*L.*zetah));
    u = -real(ifft2(1i*L.*psih));
    v =  real(ifft2(1i*K.*psih));
    rhs = -fft2( u.*zetax + v.*zetay ) - nu*Ksq.*zetah;

    % Forward Euler timestep
    zetah = zetah + dt*rhs;
    t = t + dt;
end
